### PR TITLE
Query library ifelse support for vector conditions and scalar true and false conditions and percentile bug fix

### LINKED ISSUE #2611: Numeric.percentile bug for empty input
**Description**

During the nightly docs test which runs all code snippets in the docs for validity, the following error occurred while executing this Python snippet

```python
import deephaven.perfmon as pm

qup = pm.query_update_performance(1)
qp = pm.query_performance(1)
qop = pm.query_operation_performance(1)
ss = pm.server_state()
```

```
  heduler-Concurrent-3 | .b.BarrageMessageProducer | Internal Error 'e5c7be57-7583-4e82-8a0a-43d94c913ee8' io.deephaven.UncheckedDeephavenException: Failure to execute snapshot function with unchanged clock
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1160)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1019)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(ConstructSnapshot.java:599)
          at io.deephaven.server.barrage.BarrageMessageProducer.getSnapshot(BarrageMessageProducer.java:2227)
          at io.deephaven.server.barrage.BarrageMessageProducer.updateSubscriptionsSnapshotAndPropagate(BarrageMessageProducer.java:1344)
          at io.deephaven.server.barrage.BarrageMessageProducer$UpdatePropagationJob.run(BarrageMessageProducer.java:1019)
          at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
          at java.util.concurrent.FutureTask.run(FutureTask.java:264)
          at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
          at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:160)
          at java.lang.Thread.run(Thread.java:829)
  caused by:
  io.deephaven.engine.table.impl.select.FormulaEvaluationException: In formula: UGPCycleP90Secs = divide(percentile(0.9, IntervalUGPCyclesTimeMicros), (multiply(1000, 1000.0)))
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.applyFormulaPerItem(Formula.java:161)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.lambda$fillChunkHelper$4(Formula.java:150)
          at io.deephaven.engine.rowset.RowSequence.lambda$forAllRowKeys$0(RowSequence.java:179)
          at io.deephaven.engine.rowset.impl.RowSequenceRowKeysChunkImpl.forEachRowKey(RowSequenceRowKeysChunkImpl.java:206)
          at io.deephaven.engine.rowset.RowSequence.forAllRowKeys(RowSequence.java:178)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.fillChunkHelper(Formula.java:148)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.fillChunk(Formula.java:125)
          at io.deephaven.engine.table.impl.sources.ViewColumnSource.fillChunk(ViewColumnSource.java:254)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource$FillContext.doOrderedFillAscending(RedirectedColumnSource.java:717)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.doFillChunk(RedirectedColumnSource.java:431)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.fillChunk(RedirectedColumnSource.java:409)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.getSnapshotDataAsChunkList(ConstructSnapshot.java:1462)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.serializeAllTable(ConstructSnapshot.java:1344)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.lambda$constructBackplaneSnapshotInPositionSpace$2(ConstructSnapshot.java:594)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1087)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1019)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(ConstructSnapshot.java:599)
          at io.deephaven.server.barrage.BarrageMessageProducer.getSnapshot(BarrageMessageProducer.java:2227)
          at io.deephaven.server.barrage.BarrageMessageProducer.updateSubscriptionsSnapshotAndPropagate(BarrageMessageProducer.java:1344)
          at io.deephaven.server.barrage.BarrageMessageProducer$UpdatePropagationJob.run(BarrageMessageProducer.java:1019)
          at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
          at java.util.concurrent.FutureTask.run(FutureTask.java:264)
          at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
          at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:160)
          at java.lang.Thread.run(Thread.java:829)
  caused by:
  java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 0
          at io.deephaven.function.Numeric.percentile(Numeric.java:12816)
          at io.deephaven.function.Numeric.percentile(Numeric.java:12792)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.applyFormulaPerItem(Formula.java:159)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.lambda$fillChunkHelper$4(Formula.java:150)
          at io.deephaven.engine.rowset.RowSequence.lambda$forAllRowKeys$0(RowSequence.java:179)
          at io.deephaven.engine.rowset.impl.RowSequenceRowKeysChunkImpl.forEachRowKey(RowSequenceRowKeysChunkImpl.java:206)
          at io.deephaven.engine.rowset.RowSequence.forAllRowKeys(RowSequence.java:178)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.fillChunkHelper(Formula.java:148)
          at io.deephaven.temp.cm336573424987228486v55_0.Formula.fillChunk(Formula.java:125)
          at io.deephaven.engine.table.impl.sources.ViewColumnSource.fillChunk(ViewColumnSource.java:254)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource$FillContext.doOrderedFillAscending(RedirectedColumnSource.java:717)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.doFillChunk(RedirectedColumnSource.java:431)
          at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.fillChunk(RedirectedColumnSource.java:409)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.getSnapshotDataAsChunkList(ConstructSnapshot.java:1462)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.serializeAllTable(ConstructSnapshot.java:1344)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.lambda$constructBackplaneSnapshotInPositionSpace$2(ConstructSnapshot.java:594)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1087)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1019)
          at io.deephaven.engine.table.impl.remote.ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(ConstructSnapshot.java:599)
          at io.deephaven.server.barrage.BarrageMessageProducer.getSnapshot(BarrageMessageProducer.java:2227)
          at io.deephaven.server.barrage.BarrageMessageProducer.updateSubscriptionsSnapshotAndPropagate(BarrageMessageProducer.java:1344)
          at io.deephaven.server.barrage.BarrageMessageProducer$UpdatePropagationJob.run(BarrageMessageProducer.java:1019)
          at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
          at java.util.concurrent.FutureTask.run(FutureTask.java:264)
          at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
          at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:160)
          at java.lang.Thread.run(Thread.java:829)
```

**Additional details and attachments**

When I run the code locally, it does not error. It's possible the state only happened from the docs snapshots running ~75 other examples before this and resetting the workspace just by clearing the symbols added to the python globals scope and local scope. I am unable to reproduce, but figured the out of bounds exception is an uncaught edge case in some code somewhere.

**Versions**
 - Deephaven: [Edge at this commit](https://github.com/deephaven/deephaven-core/tree/7c0a0965ce80b4570aa8cb1d065f37a44445ef05)
 - OS: Github Actions ubuntu-20.04

### LINKED ISSUE #3022: Query library `ifelse` support for vector conditions and scalar true and false conditions
Support `ifelse` functions in the query library that accept vector condition arguments and scalar true / false values.

### PR DESCRIPTION
**Pull request by chipkent**
Resolves #3022.
Resolves #2611

### CODE REVIEW

File: engine/function/src/templates/Basic.ftl [hunk 1]
```diff
@@ -786,6 +786,63 @@ public class Basic {
...
+     */
+    public static <T> T[] ifelseObj(BooleanVector condition, T trueCase, T falseCase) {
+        if (condition == null) {
+            return null;
+        }
+
+        final int n_c = condition.intSize("condition");
+        @SuppressWarnings("unchecked") final T[] result;
+
+        if ( trueCase == null && falseCase == null ) {
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:808]: If you feel like changing it (different style, but equivalent). I think it's a tad shorter and more readable.

However if you *don't* feel like it, please remove the extra spacing inside the parens for lines 808, 810, 812

```
final T typeToUse = trueCase != null ? trueCase : falseCase;
if (typeToUse == null) {
   throw new IllegalArgumentException("trueCase and falseCase are null.  Can not resolve a return type.");
}
if (trueCase != null && falseCase != null && trueCase.getClass() != falseCase.getClass()) {
    throw new IllegalArgumentException("Inputs have different types. trueCase=" + trueCase.getClass() + " falseCase=" + falseCase.getClass());
}
result = (T[])Array.newInstance(typeToUse.getClass(), n_c);
```

File: engine/function/src/templates/Basic.ftl [hunk 2]
```diff
@@ -786,6 +786,63 @@ public class Basic {
...
+        } else if ( falseCase == null ) {
+            result = (T[])Array.newInstance(trueCase.getClass(), n_c);
+        } else if (!trueCase.getClass().equals(falseCase.getClass())) {
+            throw new IllegalArgumentException("Inputs have different types. trueCase=" + trueCase.getClass() + " falseCase=" + falseCase.getClass());
+        } else {
+            result = (T[])Array.newInstance(trueCase.getClass(), n_c);
+        }
+
+        for (int i=0; i < n_c; i++) {
+            result[i] = condition.get(i) == null ? null : (condition.get(i) ? trueCase : falseCase);
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:821]: Can we put condition.get(i) into a variable so we don't call it twice?

```
final Boolean c = condition.get(i);
result[i] = c == null ? null : (c ? trueCase : falseCase);
```

File: engine/function/src/templates/Basic.ftl [hunk 3]
```diff
@@ -786,6 +786,63 @@ public class Basic {
...
+    /**
+     * Returns elements from either trueCase or falseCase, depending on condition.
+     *
+     * @param condition a boolean value used to select output values.
+     * @param trueCase value returned when condition is true.
+     * @param falseCase value returned when condition is false.
+     * @return An array of T whose values are determined by the corresponding elements of condition, trueCase, and falseCase.
+     *         The result element will be trueCase if the condition element is true;
+     *         falseCase if the condition element is false; or null if the condition element is null.
+     *         Returns null if any of the inputs is null.
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:798]: Returns null if condition is null.

File: engine/function/src/templates/Basic.ftl [hunk 4]
```diff
@@ -786,6 +786,63 @@ public class Basic {
...
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:836]: Returns null if condition is null.

File: engine/function/src/templates/Basic.ftl [hunk 5]
```diff
@@ -1585,6 +1642,52 @@ public class Basic {
...
+    /**
+     * Returns elements from either trueCase or falseCase, depending on condition.
+     *
+     * @param condition a boolean value used to select output values.
+     * @param trueCase value returned when condition is true.
+     * @param falseCase value returned when condition is false.
+     * @return An array of ${pt.primitive} whose values are determined by the corresponding elements of condition, trueCase, and falseCase.
+     *         The result element will be trueCase if the condition element is true;
+     *         falseCase if the condition element is false; or the Deephaven null constant if the condition element is null.
+     *         Returns null if any of the inputs is null.
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:1654]: Returns null if condition is null.

File: engine/function/src/templates/Basic.ftl [hunk 6]
```diff
@@ -1585,6 +1642,52 @@ public class Basic {
...
+        if (condition == null) {
+            return null;
+        }
+
+        final int n_c = condition.intSize("condition");
+
+        final ${pt.primitive}[] result = new ${pt.primitive}[n_c];
+
+        for (int i=0; i < n_c; i++) {
+            result[i] = condition.get(i) == null ? ${pt.null} : (condition.get(i) ? trueCase : falseCase);
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:1666]: same as above regarding prefer not to repeat the evaluation of `condition.get(i)`

File: engine/function/src/templates/Basic.ftl [hunk 7]
```diff
@@ -1585,6 +1642,52 @@ public class Basic {
...
+    /**
+     * Returns elements from either trueCase or falseCase, depending on condition.
+     *
+     * @param condition a boolean value used to select output values.
+     * @param trueCase value returned when condition is true.
+     * @param falseCase value returned when condition is false.
+     * @return An array of ${pt.primitive} whose values are determined by the corresponding elements of condition, trueCase, and falseCase.
+     *         The result element will be trueCase if the condition element is true;
+     *         falseCase if the condition element is false; or the Deephaven null constant if the condition element is null.
+     *         Returns null if any of the inputs is null.
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:1681]: Returns null if condition is null.

File: engine/function/src/templates/Basic.ftl [hunk 8]
```diff
@@ -1585,6 +1641,53 @@ public class Basic {
...
+        if (condition == null) {
+            return null;
+        }
+
+        final int n_c = condition.intSize("condition");
+
+        final ${pt.primitive}[] result = new ${pt.primitive}[n_c];
+
+        for (int i=0; i < n_c; i++) {
+        final Boolean c = condition.get(i);
```
Comment (kosak) [inline engine/function/src/templates/Basic.ftl:1665]: indentation for these two lines looks wrong.

**Review by kosak — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/1823
Conceptual: https://github.com/deephaven/deephaven.io/issues/1825
Reference: https://github.com/deephaven/deephaven.io/issues/1826
Blog: https://github.com/deephaven/deephaven.io/issues/1824

**Pull request merged**